From 50f0b3f23162302487d16f15b5d63aa9e955776e Mon Sep 17 00:00:00 2001
From: Luca Beldi <v.ronin@yahoo.it>
Date: Thu, 15 Apr 2021 23:41:28 +0100
Subject: [PATCH 010/132] Fix QAbstractItemModelTester false positive

When rows are removed from a model with no columns,
the test should not report a problem if indexes are invalid

Fixes: QTBUG-92886
Change-Id: I7a042dfdb2575f87208a00cbed13db3869807f84
Reviewed-by: David Faure <david.faure@kdab.com>
(cherry picked from commit fed2c0d23614df1b96dcc8746223501b07597a52)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
(cherry picked from commit cbec5527903a6ee8fc711b1a352dc17b6efada71)
---
 src/testlib/qabstractitemmodeltester.cpp                   | 4 ++--
 .../tst_qabstractitemmodeltester.cpp                       | 7 +++++--
 2 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/src/testlib/qabstractitemmodeltester.cpp b/src/testlib/qabstractitemmodeltester.cpp
index a85b291220..a2c643660a 100644
--- a/src/testlib/qabstractitemmodeltester.cpp
+++ b/src/testlib/qabstractitemmodeltester.cpp
@@ -720,12 +720,12 @@ void QAbstractItemModelTesterPrivate::rowsAboutToBeRemoved(const QModelIndex &pa
     Changing c;
     c.parent = parent;
     c.oldSize = model->rowCount(parent);
-    if (start > 0) {
+    if (start > 0 && model->columnCount(parent) > 0) {
         const QModelIndex startIndex = model->index(start - 1, 0, parent);
         MODELTESTER_VERIFY(startIndex.isValid());
         c.last = model->data(startIndex);
     }
-    if (end < c.oldSize - 1) {
+    if (end < c.oldSize - 1 && model->columnCount(parent) > 0) {
         const QModelIndex endIndex = model->index(end + 1, 0, parent);
         MODELTESTER_VERIFY(endIndex.isValid());
         c.next = model->data(endIndex);
diff --git a/tests/auto/testlib/qabstractitemmodeltester/tst_qabstractitemmodeltester.cpp b/tests/auto/testlib/qabstractitemmodeltester/tst_qabstractitemmodeltester.cpp
index 08419b335e..4aa3f8d60b 100644
--- a/tests/auto/testlib/qabstractitemmodeltester/tst_qabstractitemmodeltester.cpp
+++ b/tests/auto/testlib/qabstractitemmodeltester/tst_qabstractitemmodeltester.cpp
@@ -105,13 +105,16 @@ void tst_QAbstractItemModelTester::standardItemModel()
     model.insertColumns(0, 5, model.index(1, 3));
 }
 
-void tst_QAbstractItemModelTester::standardItemModelZeroColumns() // QTBUG-92220
+void tst_QAbstractItemModelTester::standardItemModelZeroColumns()
 {
     QStandardItemModel model;
-
     QAbstractItemModelTester t1(&model);
+    // QTBUG-92220
     model.insertRows(0, 5);
     model.removeRows(0, 5);
+    // QTBUG-92886
+    model.insertRows(0, 5);
+    model.removeRows(1, 2);
 }
 
 void tst_QAbstractItemModelTester::testInsertThroughProxy()
-- 
2.35.1

